channeldb/mp_payment.go

```diff
@@ -280,6 +305,55 @@ func (m *MPPayment) Registrable() error {
 	return nil
 }
 
+// setState creates and attaches a new MPPaymentState to the payment. It also
+// updates the payment's status based on its current state.
+func (m *MPPayment) setState() error {
+	// Fetch the total amount and fees that has already been sent in
+	// settled and still in-flight shards.
+	sentAmt, fees := m.SentAmt()
+
+	// Sanity check we haven't sent a value larger than the payment amount.
+	totalAmt := m.Info.Value
+	if sentAmt > totalAmt {
+		return fmt.Errorf("%w: sent=%v, total=%v", ErrSentExceedsTotal,
+			sentAmt, totalAmt)
+	}
+
+	// Get any terminal info for this payment.
+	settle, failure := m.TerminalInfo()
+
+	// If either an HTLC settled, or the payment has a payment level
+	// failure recorded, it means we should terminate the moment all shards
+	// have returned with a result.
+	terminate := settle != nil || failure != nil
+
+	// Now determine the payment's status.
+	status, err := decidePaymentStatus(m.HTLCs, m.FailureReason)
+	if err != nil {
+		return err
+	}
+
+	// Update the payment state and status.
+	m.State = &MPPaymentState{
+		NumAttemptsInFlight: len(m.InFlightHTLCs()),
```

Comment:
Exactly. We need to snapshot the current payment status because the on-disk payment is being read/written by other goroutines.

Reply:
it's indeed part of the `fetchPayment`, and `setState` is used there to set the state. The `MPPaymentState` is used to wrap the payment state that's relevant to the payment lifecycle machine, which makes it easier to mock the state in tests. For instance, instead of mocking `InFlightHTLCs`, `SentAmt`, `TerminalInfo`, and `decidePaymentStatus`, I can mock `GetState`, and drive the state change there in payment lifecycle unit tests.